File: LeetCode/backtracking/39.combination-sum.py
#
# @lc app=leetcode id=39 lang=python3
#
# [39] Combination Sum
#

# @lc code=start
class Solution:
    result = []
    target = 0
    candidates = []
    def backtrack(self, trackSum, subset, start):
        if trackSum == self.target:
            self.result.append(subset.copy())
        for idx in range(start, len(self.candidates)):
            trackSum += self.candidates[idx]
            if trackSum <= self.target:
                subset.append(self.candidates[idx])
                self.backtrack(trackSum, subset, idx)
                subset.pop(-1)
            trackSum -= self.candidates[idx]

        pass
    def combinationSum(self, candidates: List[int], target: int) -> List[List[int]]:
        self.result = []
        self.candidates = candidates
        self.candidates.sort()
        self.target = target
        self.backtrack(trackSum = 0, subset=[], start = 0)
        return self.result
        
# @lc code=end

File: LeetCode_Go/Backtrack/39.combination-sum.go
/*
 * @lc app=leetcode id=39 lang=golang
 *
 * [39] Combination Sum
 */

// @lc code=start

// Given an array of distinct integers candidates and a target integer target, return a list of all unique combinations of candidates where the chosen numbers sum to target. You may return the combinations in any order. The same number may be chosen from candidates an unlimited number of times.

// backtracking(startIdx:), after a choice is made, we can still make the same choice unlimit times, so we call backtracking(startIdx:) instead of backtracking(startIdx+1:). Exit condition: sum >= target.

func combinationSum(candidates []int, target int) [][]int {
	ans := [][]int{}
	backtracking(0, 0, target, []int{}, candidates, &ans)
	return ans
}

func backtracking(start, sum, target int, choices, candidates []int, ans *[][]int) {
	if sum >= target {
		if sum == target {
			// save to ans
			temp := make([]int, len(choices))
			copy(temp, choices)
			*ans = append(*ans, temp)
		}
		return
	}

	for i := start; i < len(candidates); i++ {
		// do choice
		choices = append(choices, candidates[i])
		sum += candidates[i]
		backtracking(i, sum, target, choices, candidates, ans)
		// undo choice
		choices = choices[:len(choices)-1]
		sum -= candidates[i]
	}
}

// @lc code=end

